pkg/sql/rowexec/kv_scan.go

```diff
@@ -0,0 +1,176 @@
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
+	"github.com/cockroachdb/cockroach/pkg/sql/rowenc"
+	"github.com/cockroachdb/cockroach/pkg/sql/rowinfra"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+)
+
+// kvScan is used read raw KV values in a computation flow for purposes,
+// of mainly implementing crdb_internal.scan. This execinfra.RowSource
+// only returns a row format of key, value, and ts.
+type kvScan struct {
+	execinfra.ProcessorBase
+	spans           []roachpb.Span
+	rowIdx          int
+	bytesRead       int64
+	rowsRead        int64
+	scanResponses   []kvpb.ResponseUnion
+	currentResponse int
+	row             rowenc.EncDatumRow
+}
+
+// Start is part of the execinfra.RowSource interface.
+func (ks *kvScan) Start(ctx context.Context) {
+	ctx = ks.StartInternal(ctx, "index-scan")
+	// Set up the first batch.
+	ba := ks.EvalCtx.Txn.NewBatch()
+	for spanIdx := range ks.spans {
+		scan := kvpb.NewScan(ks.spans[spanIdx].Key,
+			ks.spans[spanIdx].EndKey,
+			false)
+		ba.AddRawRequest(scan)
+	}
+	ba.Header.MaxSpanRequestKeys = int64(rowinfra.ProductionKVBatchSize)
+	ba.Header.TargetBytes = int64(rowinfra.GetDefaultBatchBytesLimit(false))
+	err := ks.EvalCtx.Txn.Run(ctx, ba)
+	if err != nil {
+		ks.MoveToDraining(err)
+	}
+	ks.scanResponses = ba.RawResponse().Responses
+	// Set up the results for the rows.
+	ks.row = make(rowenc.EncDatumRow, 3)
+}
+
+// getNextScan will issue any additional scans to read additional data,
+// using any response spans.
+func (ks *kvScan) getNextScan(ctx context.Context) error {
+	// Generate a new batch with all of the resume spans
+	// from our first request.
+	ba := ks.EvalCtx.Txn.NewBatch()
+	for i := ks.currentResponse; i < len(ks.scanResponses); i++ {
+		scan := ks.scanResponses[i].GetScan()
+		nextSpan := scan.ResumeSpan
+		scanRequest := kvpb.NewScan(nextSpan.Key,
+			nextSpan.EndKey,
+			false /*forUpdate*/)
+		ba.AddRawRequest(scanRequest)
+	}
+	ba.Header.MaxSpanRequestKeys = int64(rowinfra.ProductionKVBatchSize)
+	err := ks.EvalCtx.Txn.Run(ctx, ba)
+	if err != nil {
+		return err
+	}
+	ks.scanResponses = ba.RawResponse().Responses
+	ks.currentResponse = 0
+	ks.rowIdx = 0
+	return nil
+}
+
+// Next is part of the execinfra.RowSource interface.
+func (ks *kvScan) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerMetadata) {
+	if ks.State != execinfra.StateRunning {
+		return nil, ks.DrainHelper()
+	}
+	moreRows := true
+	var err error
+	for moreRows {
+		currentScan := ks.scanResponses[ks.currentResponse].GetScan()
+		// The current scan is out of data, so check if we have more
+		// data.
+		if ks.rowIdx >= len(currentScan.Rows) {
+			// If there is a resume span, then we need to
+			// make another request to continue.
+			if currentScan.ResumeSpan != nil {
+				err = ks.getNextScan(ks.Ctx())
```

Comment:
I'm reviewing this pre-coffee, so I might have confused myself, but the current code here seems like it heavily depends on DistSender not our scan requests in parallel (which I think will be true because we set MaxSpanRequestKeys), but it makes me a little nervous and perhaps deserves a comment.  Feel free to ignore if this is more obvious to others or I'm just confused.

For example, say we sent our first batch request and we had two scan requests for the spans: a-d, f-k. And say that f-k finished completely, but a-d had a resume span, then in the current code would we ever read the results of f-k?